master/buildbot/reporters/notifier.py

```diff
@@ -146,160 +158,37 @@ def stopService(self):
             yield self._workerMissingConsumer.stopConsuming()
             self._workerMissingConsumer = None
 
-    def wantPreviousBuild(self):
-        return "change" in self.mode or "problem" in self.mode
-
     @defer.inlineCallbacks
-    def buildsetComplete(self, key, msg):
-        if not self.buildSetSummary:
-            return
-        bsid = msg['bsid']
-        res = yield utils.getDetailsForBuildset(
-            self.master, bsid,
-            wantProperties=self.messageFormatter.wantProperties,
-            wantSteps=self.messageFormatter.wantSteps,
-            wantPreviousBuild=self.wantPreviousBuild(),
-            wantLogs=self.messageFormatter.wantLogs)
-
-        builds = res['builds']
-        buildset = res['buildset']
+    def _send_reports_for_type(self, type, key, msg):
```

Comment:
I don't think the switch based on type of generator is optimal API. I see a bit of dependency inversion here.

I think it would be better to let the generator advertise what kind of event it is interested in, and then just have generic reconfigureService which would register to the right events. 

This makes the whole thing a bit more complex and would make you have the generator being child services.
We have to take care of unregistering for events when a generator is not configured anymore (via buildbot reconfig)

Reply:
making generators child services would be the best in term of separation of concerns, but I admit this is a bit complicated to handle the life cycle of objects.
So for this PR, lets manage the event register in the notifier. 
Note that this effectively makes the notifier another event router.

Reply:
I agree that it would be better to request events in the generators. This PR was just the first bite into the problem, I was focusing into handling of generator output and wanted to do as few other changes as possible. We still ended up with a PR with more than a thousand lines changed.